crates/gpui/src/platform/linux/wayland/cursor.rs

```diff
@@ -0,0 +1,48 @@
+use wayland_client::{Connection, QueueHandle};
+use wayland_client::protocol::wl_compositor::WlCompositor;
+use wayland_client::protocol::wl_pointer::WlPointer;
+use wayland_client::protocol::wl_shm::WlShm;
+use wayland_client::protocol::wl_surface::WlSurface;
+use wayland_cursor::{CursorImageBuffer, CursorTheme};
+use crate::platform::linux::wayland::WaylandClientState;
+
+pub(crate) struct Cursor {
+    theme: CursorTheme,
+    current_icon_name: String,
+    surface: WlSurface,
+    serial_id: u32,
+}
+
+impl Cursor {
+    pub fn new(connection: &Connection, compositor: &WlCompositor, qh: &QueueHandle<WaylandClientState>, shm: &WlShm, size: u32) -> Self {
+        let mut cursor_theme = CursorTheme::load(&connection, shm.clone(), size)
+            .expect("Could not load cursor theme");
+        Self {
+            theme: cursor_theme,
+            current_icon_name: "".to_string(),
+            surface: compositor.create_surface(qh, ()),
+            serial_id: 0,
+        }
+    }
+
+    pub fn set_serial_id(&mut self, serial_id: u32) {
+        self.serial_id = serial_id;
+    }
+
+    pub fn set_icon(&mut self, wl_pointer: &WlPointer, cursor_icon_name: String) {
+        if self.current_icon_name != cursor_icon_name {
+            let cursor = self.theme.get_cursor(cursor_icon_name.as_str()).expect("Cursor not provided by theme");
```

Comment:
Please don't use hard panics like this. It would be best to log an error then provide some sort of fallback/failsafe. I think the `.expect()` above this one is probably okay but maybe a safe default would help?.

I wrapped this use in an `if let Some(cursor)` myself.

Reply:
Right. Fixed.